pennylane/_qubit_device.py

```diff
@@ -136,6 +141,33 @@ def __init__(self, wires=1, shots=None, cache=0, analytic=None):
         """OrderedDict[int: Any]: Mapping from hashes of the circuit to results of executing the
         device."""
 
+        self.custom_decomps = custom_decomps
+
+        if custom_decomps:
+            # Get a list of all the operations requiring custom decomps
+            custom_op_names = [
+                op if isinstance(op, str) else op.__name__ for op in custom_decomps.keys()
+            ]
+            custom_decomp_condition = lambda obj: obj.name in custom_op_names
+
+            # Combined rule for expansion; our custom rule plus regular device conditions
+            stop_at = qml.BooleanFn(custom_decomp_condition) and self.stopping_condition
+
+            # Create and set the custom expand function for this device
+            self.custom_fn = qml.transforms.create_expand_fn(
+                depth=10,
+                stop_at=stop_at,
+                device=self
+            )
+
+            # Create a new custom expand function that runs the decomposition
+            # within the context.
+            def custom_decomp_expand(self, circuit, max_expansion=10):
+                with self.custom_decomp_context():
+                    return self.custom_fn(circuit, max_expansion)
```

Comment:
The `with` statement does seem to alter the decomposition; if you do something like `print(qml.CNOT.decomposition(wires=[0, 1])` with the new decomposition, it will show up correctly as the custom function, however when the custom expand function is later called, the decomposition is not overridden.